pyam/iiasa.py

```diff
@@ -17,51 +19,110 @@
 # quiet this fool
 logging.getLogger('requests').setLevel(logging.WARNING)
 
-_URL_TEMPLATE = 'https://db1.ene.iiasa.ac.at/{}-api/rest/v2.1/'
-_AUTH_URL = 'https://db1.ene.iiasa.ac.at/EneAuth/config/v1/anonym'
+_BASE_URL = 'https://db1.ene.iiasa.ac.at/EneAuth/config/v1'
+_CITE_MSG = """
+You are connected to the {} scenario explorer hosted by IIASA.
+ If you use this data in any published format, please cite the
+ data as provided in the explorer guidelines: {}.
+""".replace('\n', '')
 
-_CITATIONS = {
-    'iamc15': 'D. Huppmann, E. Kriegler, V. Krey, K. Riahi, '
-    'J. Rogelj, S.K. Rose, J. Weyant, et al., '
-    'IAMC 1.5C Scenario Explorer and Data hosted by IIASA. '
-    'IIASA & IAMC, 2018. '
-    'doi: 10.22022/SR15/08-2018.15429, '
-    'url: data.ene.iiasa.ac.at/iamc-1.5c-explorer'
-}
 
+def _get_token(creds):
+    # get anonymous auth if creds
+    if creds is None:  # anonymously
+        url = '/'.join([_BASE_URL, 'anonym'])
+        return requests.get(url).json()
 
-def valid_connection_names():
-    return list(_CITATIONS.keys())
+    # otherwise read creds and try to login
+    try:
+        if isinstance(creds, collections.Mapping):
+            user, pw = creds['username'], creds['password']
+        else:
+            user, pw = creds
+    except Exception as e:
+        msg = 'Could not read credentials: {}\n{}'.format(
+            creds, str(e))
+        raise type(e)(msg)
+
+    headers = {'Accept': 'application/json',
+               'Content-Type': 'application/json'}
+    data = {'username': user, 'password': pw}
+    url = '/'.join([_BASE_URL, 'login'])
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+    if not response.ok:  # auth failed, provide user a nice message
+        raise RuntimeError('Login failed for user: {}'.format(user))
+
+    return response.json()
 
 
 class Connection(object):
     """A class to facilitate querying an IIASA scenario explorer database"""
 
-    def __init__(self, name):
+    def __init__(self, name=None, creds=None):
         """
         Parameters
         ----------
-        name : str
+        name : str, optional
             A valid database name. For available options, see
             valid_connection_names().
+        creds : list-like or dict, optional
+            An ordered container with entries of 'username' and 'password',
+            or a dictionary with the same keys.
         """
-        valid = valid_connection_names()
+        self._token = _get_token(creds)
+
+        # connect if provided a name
+        self._connected = None
+        if name:
+            self.connect(name)
+
+    @property
+    @lru_cache()
+    def valid_connections(self):
+        url = '/'.join([_BASE_URL, 'applications'])
+        headers = {'Authorization': 'Bearer {}'.format(self._token)}
+        response = requests.get(url, headers=headers).json()
```

Comment:
would it be better to check response status here?

Reply:
Yes, probably so. I can write a convenience function to do this for all requests.